### ISSUE TITLE
Temporarily save the breakpoints in the frontend

### ISSUE THREAD

**Issue by JohanMabille** [jupyterlab/jupyterlab#10090]
### Problem

Some users would like to be able to do the following:
- start the debugger
- set some breakpoints
- restart and run all cells
- have the execution stopping on the first breakpoint.

With the current implementation, the list of breakpoints is lost when the kernel restarts.

### Proposed Solution

A first solution could be to store the breakpoints as part of the notebook metadata. This would allow to keep the breakpoints between different debug seesions, even if we close Jupyter Lab or refresh the page. However, storing the breakpoints in the kernel (the current behavior) would become useless and we may not want to keep the breakpoints between two Lab sessions.

Another solution would be to keep the breakpoints in memory, even when the debugger is stopped or the kernel is restarted. This would allow to keep the breakpoints through different kernel restarts, but the kernel would remain responsible for storing and sending back the breakpoints to the frontend.

### Additional context

The second solution would be similar to the behavior of Visual Studio which keeps the breakpoints between two debug sessions but lost them when you close it.

**Comment by martinRenou** [comment]
> Another solution would be to keep the breakpoints in memory, even when the debugger is stopped or the kernel is restarted. This would allow to keep the breakpoints through different kernel restarts, but the kernel would remain responsible for storing and sending back the breakpoints to the frontend.

Do you think there would be a possible race condition when doing "Restart and run all cells"? We need to be sure that the breakpoints are received by the kernel and that the debugger has started before running the first cell.

**Comment by JohanMabille** [comment]
Indeed, that would require either to change the code of the restart so that it depends on the debugger, or to implmeent a custom "restart and debug" command.

**Comment by jtpio** [comment]
Could this be implemented as third-party extension, so it can be experimented with? Or do you think it should part of the core debugger experience?

**Comment by jtpio** [comment]
For reference this is where the "Restart and Run All" logic is currently implemented:

https://github.com/jupyterlab/jupyterlab/blob/97fd1a19bb5fcaa718b43a5d216ba9bf402d3af9/packages/notebook-extension/src/index.ts#L1581-L1600

And also here, which seems to be a duplication of the same logic:

https://github.com/jupyterlab/jupyterlab/blob/97fd1a19bb5fcaa718b43a5d216ba9bf402d3af9/packages/notebook-extension/src/index.ts#L2466-L2476

**Comment by jtpio** [comment]
> Another solution would be to keep the breakpoints in memory, even when the debugger is stopped or the kernel is restarted. This would allow to keep the breakpoints through different kernel restarts, but the kernel would remain responsible for storing and sending back the breakpoints to the frontend.

Maybe that could be the way to go actually? And would solve other usability issues with the debugger such as being able to keep the breakpoints when manually turning off and on (not just restarts), at least temporarily while still on the page (breakpoints would still be lost after reloading the page). 

This could also be made a settings so end users could decide whether or not to keep breakpoints in memory.

At the moment turning on and off the debugger shows and hides the gutters, loosing the breakpoints if there were any:

https://user-images.githubusercontent.com/591645/121672146-bf26a300-caaf-11eb-8802-51f6d5a63c3f.mp4

So far the debugger frontend has been pretty much stateless, so we should also check what this would imply if we now decide to start keeping state on the frontend.

---

Not sure the other way consisting in replacing the behavior of the restart and run all command would work well. 

It would require adding `@jupyterlab/debugger` as a dependency of `@jupyterlab/notebook-extension`, although it could of course be treated as optional. And bring some debugger logic into the notebook extension?

**Comment by martinRenou** [comment]
It would also be nice to have a more "dynamic" UX, where breakpoints can always be added. If a first breakpoint is added by the user and the debugger was not started, the debugger would start automatically.

This would make the workflow "start debugger" > "add breakpoints" > "run" less cumbersome

**Comment by martinRenou** [comment]
It looks like I have memory issues 😅 I forgot I opened an issue for that, sorry

**Comment by krassowski** [comment]
> Or do you think it should part of the core debugger experience?

Speaking as a user I was very surprised to see that this removed my break after trying to use debugger. I don't want "restart and debug", I prefer to have "restart and run all" and "restart and run up to selected cell" work well with debugger (I have those mapped to keyboard shortcuts and a new command would only confuse me). Also, while switching line numbers on/off is already discussed elsewhere I just wanted to highlight that this is a pain in larger notebooks due to CodeMirror redraws causing a freeze and while I can tolerate it when I manually turn the debugger on/off, it was a bit annoying when it happened after "restart and run up to selected cell".

Important thing to consider is that this comes up very handy when the kernel **dies**. This is the reason I installed RC for ipykernel today as I am trying to debug some issues which leads to kernel dying. "Restart and run up to selected cell"/"restart all" is a prerequisite in this very case when the debugger is actually the most desired thing we can offer for the user (as the "kernel died" message is often a source of frustration and difficult to diagnose).

**Comment by JohanMabille** [comment]
> Not sure the other way consisting in replacing the behavior of the restart and run all command would work well.
>
> It would require adding @jupyterlab/debugger as a dependency of @jupyterlab/notebook-extension, although it could of course be treated as optional. And bring some debugger logic into the notebook extension?

After more thought on this, I think you're right, we should not introduce debugger logics into the notebook package. I think this logic should be part of the debugger package itself, since breakpoints are still in memory when we restart a kernel. We should also consider adding a new menu specific to the debugger, like you can find in traditional IDEs. It would make it easier for the user to toggle/remove all breakpoints, set global exceptoin breakpoint, and so on. But this can be discussed in a dedicagted issue.
